tests/abi/tools/revng-check-compatibility-with-abi/ABIRuntimeTestResultParser.cpp

```diff
@@ -11,25 +11,101 @@
+    for (auto I = ResultRegisters.batch_insert();
+         auto Reg : model::Architecture::registers(Architecture)) {
+      constexpr auto PoN = model::PrimitiveTypeKind::PointerOrNumber;
+      if (model::Register::primitiveKind(Reg) != PoN) {
+        // Only support generic registers for now.
+        // TODO: add vector register support.
+        continue;
+      }
+
+      auto Iterator = llvm::find_if(Registers, [Reg](const RawWord &Raw) {
+        return Raw.Name == getRegisterName(Reg);
+      });
+      if (Iterator == Registers.end()) {
+        std::string Error = "Unable to find '" + getRegisterName(Reg).str()
+                            + "' in the artifact.\n";
+        revng_abort(Error.c_str());
+      }
+      I.emplace(std::move(*Iterator), Architecture);
+    }
 
-using namespace abi::artifact;
+    return State(std::move(ResultRegisters), std::move(Stack));
+  }
+};
 
-struct SingleRun {
+struct SingleIteration {
+  // Basic information.
   llvm::StringRef Function;
-  Registers Registers;
-  Stack Stack;
-  Arguments ReturnValue;
-  Arguments Arguments;
+  std::uint64_t Index;
+
+  // General state.
+  RawState StateBeforeTheCall;
+  RawState StateAfterTheCall;
+  RawState StateAfterTheReturn;
+
+  // Test specific.
+  std::vector<RawWord> ReturnValueAddress; // Return value tests only.
+  std::vector<Argument> ReturnValue; // Return value tests only.
+  std::vector<Argument> ExpectedReturnValue; // Return value tests only.
+  std::vector<Argument> Arguments; // Argument tests only.
+
+  bool isArgumentTest() const {
+    return !Arguments.empty() && ReturnValueAddress.empty()
+           && ReturnValue.empty() && ExpectedReturnValue.empty();
+  }
+
+  bool isReturnValueTest() const {
+    return Arguments.empty() && !ReturnValueAddress.empty()
+           && !ReturnValue.empty() && !ExpectedReturnValue.empty();
+  }
+
+  TestState extractTheState(model::Architecture::Values Arch) && {
+    return TestState{
+      .StateBeforeTheCall = std::move(StateBeforeTheCall).extract(Arch),
+      .StateAfterTheCall = std::move(StateAfterTheCall).extract(Arch),
+      .StateAfterTheReturn = std::move(StateAfterTheReturn).extract(Arch),
+    };
+  }
+
+  std::variant<ArgumentTest, ReturnValueTest>
+  extract(model::Architecture::Values Architecture) && {
+    if (isArgumentTest()) {
+      revng_assert(!isReturnValueTest());
+      return ArgumentTest(std::move(*this).extractTheState(Architecture),
+                          std::move(Arguments));
+    } else {
+      revng_assert(isReturnValueTest());
+      revng_assert(ReturnValueAddress.size() == 1);
+      revng_assert(ReturnValue.size() == 1);
+      revng_assert(ExpectedReturnValue.size() == 1);
+      return ReturnValueTest(std::move(*this).extractTheState(Architecture),
+                             std::move(ReturnValueAddress[0]),
+                             std::move(ReturnValue[0]),
+                             std::move(ExpectedReturnValue[0]));
```

Comment:
Drop all the `std::move`, they make the code unreadable and obscure.
Drop them all.

Reply:
I simplified the data structures here a lot. Please take a fresh look when you have a moment.